test/e2e/storage/testsuites/provisioning.go

```diff
@@ -297,12 +299,73 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
 		framework.ExpectNoError(err)
 		ginkgo.DeferCleanup(f.DeleteNamespace, valNamespace.Name)
 
+		crdManifestPath := "test/e2e/testing-manifests/storage-csi/any-volume-datasource/crd/populator.storage.k8s.io_volumepopulators.yaml"
+		crdItems, err := storageutils.LoadFromManifests(crdManifestPath)
+		framework.ExpectNoError(err, "Failed to load VolumePopulator CRD manifest")
+		gomega.Expect(crdItems).To(gomega.HaveLen(1), "Expected exactly one CRD in manifest")
+
+		crd, ok := crdItems[0].(*apiextensionsv1.CustomResourceDefinition)
+		gomega.Expect(ok).To(gomega.BeTrue(), fmt.Sprintf("Resource in loaded manifest file is not a CustomResourceDefinition: %s", crdManifestPath))
+
+		config, err := framework.LoadConfig()
+		framework.ExpectNoError(err)
+		apiExtensionClient, err := crdclientset.NewForConfig(config)
+		framework.ExpectNoError(err)
+
+		var crdCreatedByTest bool
+
+		ginkgo.By(fmt.Sprintf("Checking if %s CRD exists", crd.Name))
+		_, err = apiExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Get(ctx, crd.Name, metav1.GetOptions{})
+		if err != nil && apierrors.IsNotFound(err) {
+			ginkgo.By("VolumePopulator CRD not found, test will create it and remove when done")
+			crdCreatedByTest = true
+		} else if err != nil {
+			framework.ExpectNoError(err, "Error checking for VolumePopulator CRD existence")
+		} else {
+			ginkgo.By(fmt.Sprintf("%s CRD already exists, leaving unchanged", crd.Name))
+			crdCreatedByTest = false
+		}
+
 		ginkgo.By("Deploying validator")
 		valManifests := []string{
-			"test/e2e/testing-manifests/storage-csi/any-volume-datasource/crd/populator.storage.k8s.io_volumepopulators.yaml",
 			"test/e2e/testing-manifests/storage-csi/any-volume-datasource/volume-data-source-validator/rbac-data-source-validator.yaml",
 			"test/e2e/testing-manifests/storage-csi/any-volume-datasource/volume-data-source-validator/setup-data-source-validator.yaml",
 		}
+
+		if crdCreatedByTest {
+			ginkgo.By("Creating VolumePopulator CRD")
+			err = storageutils.CreateFromManifests(ctx, f, valNamespace,
+				func(item interface{}) error { return nil },
+				crdManifestPath)
+			framework.ExpectNoError(err)
+
+			ginkgo.DeferCleanup(func(ctx context.Context) {
+				ginkgo.By(fmt.Sprintf("Cleaning up %s CRD", crd.Name))
+				err := apiExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Delete(ctx, crd.Name, metav1.DeleteOptions{})
+				if err != nil && !apierrors.IsNotFound(err) {
+					framework.Logf("Failed to delete %s CRD: %v", crd.Name, err)
+				}
+			})
```

Comment:
I would probably move this `DeferCleanup` block directly into the "CRD creation succeeded" block instead of using the `crdCreatedByTest` boolean.

Reply:
Removed - applied suggestion with using cleanup from `storageutils.CreateFromManifests`.